File: crates/yryvu-bridge/src/commands/hooks.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Tauri commands for Git hooks (issue #192).

use std::path::PathBuf;

use tauri::AppHandle;
use tauri_plugin_opener::OpenerExt;

use crate::repo::hooks::{
    hook_script_path, list_hooks as list_hooks_impl, set_hook_enabled as set_hook_enabled_impl,
    HookEntry,
};

#[tauri::command]
pub async fn list_hooks(repo_path: String) -> Result<Vec<HookEntry>, String> {
    tauri::async_runtime::spawn_blocking(move || list_hooks_impl(&PathBuf::from(repo_path)))
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn set_hook_enabled(
    repo_path: String,
    name: String,
    enabled: bool,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        set_hook_enabled_impl(&PathBuf::from(repo_path), &name, enabled)
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())
}

/// Resolve the on-disk script path for `name` (active or disabled) and
/// hand it to the OS default opener. The user's preferred editor for
/// shell scripts handles the rest — yryvu doesn't ship its own editor.
#[tauri::command]
pub async fn open_hook_script(
    app: AppHandle,
    repo_path: String,
    name: String,
) -> Result<(), String> {
    let path: PathBuf = tauri::async_runtime::spawn_blocking(move || {
        hook_script_path(&PathBuf::from(repo_path), &name)
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())?;
    app.opener()
        .open_path(path.to_string_lossy().to_string(), None::<&str>)
        .map_err(|e| e.to_string())
}

File: apps/yryvu-app/src/ipc/hooks.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

/// Mirrors `yryvu_bridge::repo::hooks::HookEntry`. The `path` field is
/// the absolute on-disk path, already accounting for the `.disabled`
/// suffix when applicable — the frontend doesn't need to recompute it.
export interface HookEntry {
  name: string;
  enabled: boolean;
  path: string;
}

/// List the hooks in the active hooks directory for a repo. Resolves
/// `core.hooksPath` first; falls back to `<repo>/.git/hooks`. Empty
/// vec when the directory doesn't exist.
export function listHooks(repoPath: string): Promise<HookEntry[]> {
  return invoke<HookEntry[]>("list_hooks", { repoPath });
}

/// Toggle a hook by renaming the file to/from the `.disabled` suffix.
/// Idempotent — calling with the current state is a no-op. Errors with
/// NotFound when neither the active nor `.disabled` file exists.
export function setHookEnabled(
  repoPath: string,
  name: string,
  enabled: boolean,
): Promise<void> {
  return invoke<void>("set_hook_enabled", { repoPath, name, enabled });
}

/// Open a hook script with the OS default app for shell scripts. The
/// backend resolves the on-disk file (active or `.disabled`) before
/// delegating to `tauri-plugin-opener`.
export function openHookScript(repoPath: string, name: string): Promise<void> {
  return invoke<void>("open_hook_script", { repoPath, name });
}
